include/deal.II/base/tensor.h

```diff
@@ -1986,10 +1962,11 @@ inline DEAL_II_CONSTEXPR DEAL_II_ALWAYS_INLINE
                          schur_product(const Tensor<rank, dim, Number> &     src1,
                                        const Tensor<rank, dim, OtherNumber> &src2)
 {
-  Tensor<rank, dim, typename ProductType<Number, OtherNumber>::type> tmp(src1);
+  Tensor<rank, dim, typename ProductType<Number, OtherNumber>::type> tmp;
 
   for (unsigned int i = 0; i < dim; ++i)
-    tmp[i] = schur_product(src1[i], src2[i]);
+    tmp[i] = schur_product(Tensor<rank - 1, dim, Number>(src1[i]),
+                           Tensor<rank - 1, dim, OtherNumber>(src2[i]));
```

Comment:
Why is the cast necessary? How is this not already the correct type?

Reply:
The problem is that `Tensor<1, dim, Number>::operator[]` returns a plain `double` instead of a `Tensor<0, dim, Number>::operator[]`.

In the original version the compiler was unable to deduce that it should use the `rank=0` template specialization in the `for` loop in  `schur_product` because it was called with two doubles once it reached `rank=1` in the recursion. By explicitly "creating" a temporary `Tensor<rank -1, dim, Number>` we avoid that problem. I actually checked that the assembly produced by the compiler in debug mode is unaffected by that change.

This probably also epxlains why we had a `rank=1` overload in the first place (because that fixes the recursion problem as well).

Reply:
Yes, `Tensor` is so lean that the compiler can disassociate/scalarize the whole thing. Think about the mess this would be at the assembly level if `dim` was a run-time variable...